src/Type/Php/ArrayKeysFunctionDynamicReturnTypeExtension.php

```diff
@@ -29,13 +29,12 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 		$arrayArg = $functionCall->getArgs()[0]->value ?? null;
 		if ($arrayArg !== null) {
 			$valueType = $scope->getType($arrayArg);
-			if ($valueType->isArray()->yes()) {
-				$array = $valueType->getKeysArray();
-				if ($valueType->isIterableAtLeastOnce()->yes()) {
-					$array = TypeCombinator::intersect($array, new NonEmptyArrayType());
-				}
-				return $array;
+
+			$array = $valueType->getKeysArray();
+			if ($valueType->isArray()->yes() && $valueType->isIterableAtLeastOnce()->yes()) {
+				$array = TypeCombinator::intersect($array, new NonEmptyArrayType());
 			}
+			return $array;
```

Comment:
This should already work just by returning `$valueType->getKeysArray();`.

Because `getKeysArray` looks correctly implemented in all ArrayType, IntersectionType, NonEmptyArrayType.

Reply:
It needs still special handling in the HasOffset* types AFAIK, see https://github.com/phpstan/phpstan-src/pull/1852#discussion_r996412603

Basically they'd need to return non-empty-array and I wasn't sure if that really makes sense. But then it should work 😊

Reply:
Yeah, I think it does make sense :)

Reply:
fixed